### PR TITLE
New Component: File Input

### PR DESCRIPTION
**Pull request by c-frater**
As per #25

### CODE REVIEW

File: assets/hexagon-icons(icomoon).json [hunk 1]
```diff
@@ -4,6 +4,32 @@
 		{
```
Comment (c-frater) [inline assets/hexagon-icons(icomoon).json]: Added the 'upload' icon to the hexagon icons fonts

File: modules/file-input/main/index.coffee [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    if resolvedOptions.multiple and not options.buttonText
+      resolvedOptions.buttonText = 'Choose Files'
+
+    selection = hx.select @selector
+      .classed 'hx-file-input', true
+
+    input = hx.detached 'input'
+      .class 'hx-file-input-hidden'
+      .attr 'type', 'file'
+      .attr 'accept', resolvedOptions.accept
```
Comment (c-frater) [inline modules/file-input/main/index.coffee]: We add this but there doesn't seem to be a consistent way of checking validity? (this is an issue when dragging/dropping files as we can't validate)

File: modules/file-input/main/index.coffee [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        e.preventDefault()
+        e.stopPropagation()
+
+      drop = (e) =>
+        preventDefault(e)
+        unless @_.disabled
+          files = e.dataTransfer.files
+          validates = true
+          for file in files
+            break if not validates
```
Comment (c-frater) [inline modules/file-input/main/index.coffee]: This validation loop currently does nothing - I'm investigating how to do this whilst retaining the current 'accepts' property of the file input

File: modules/file-input/main/index.coffee [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        .add(button)
+        .add(selectedFiles))
+      .add(dropdownDiv)
+
+  disabled: (disabled) ->
+    if arguments.length
+      disabledAttr = if disabled then 'disabled' else undefined
+      @_.input.attr 'disabled', disabledAttr
+      @_.button.attr 'disabled', disabledAttr
+      @_.disabled = disabled
```
Comment (c-frater) [inline modules/file-input/main/index.coffee]: Need to add a class here to make the 'no files' text greyed out
Could also remove the file (if set) but not sure that's necessary?

File: modules/file-input/main/index.coffee [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      this
+    else
+      @_.disabled or false
+
+  value: -> @_.fileMap.values()
+
+
```
Comment (c-frater) [inline modules/file-input/main/index.coffee]: Could do with a 'clear' or 'reset' method for programatically removing the value from the field.

File: modules/file-input/main/index.coffee [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    if arguments.length
+      disabledAttr = if disabled then 'disabled' else undefined
+      @_.input.attr 'disabled', disabledAttr
+      @_.button.attr 'disabled', disabledAttr
+      @_.disabled = disabled
+      this
+    else
+      @_.disabled or false
+
+  value: -> @_.fileMap.values()
```
Comment (c-frater) [inline modules/file-input/main/index.coffee]: As per security requirements, it is not possible to set the value of a file input so we have not exposed that functionality.

File: modules/file-input/module.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "dependencies": [
+    "component",
+    "selection",
+    "event-emitter"
```
Comment (c-frater) [inline modules/file-input/module.json]: Need to add dependencies here for Map/Util (and any others that were missed)

**Comment by c-frater** [comment]
This is currently a bit WIP but I thought I'd push it to get feedback on things I can change whilst polishing it off.
I've added comments where there are things to do, I'm sure there are things I've missed/could do better so feel free to comment/criticise :)

File: modules/file-input/main/index.coffee (hunk 6 above)
Comment (georgesimms) [inline modules/file-input/main/index.coffee]: What will happen if someone tries ? Console warning?

**Comment by georgesimms** [comment]
Screenshot?

**Comment by georgesimms** [comment]
Also, this may be a more general issue, but there is copypasta  in the disabled methods - is this avoidable ?

File: modules/file-input/main/index.coffee (hunk 6 above)
Comment (c-frater) [inline modules/file-input/main/index.coffee]: It returns the value of the file input as if they hadn't passed in a value. 
We could add a console warning but I didn't see the point warning the user that they've tried to do something that isn't documented.

**Comment by c-frater** [comment]
From top to bottom:
- Default file input, nothing entered
- 'multiple' file input, nothing entered
- Default file input with file added
- 'multiple' file input with values entered and dropdown open (dropdown is only shown when there is more than 1 file)

file-input

File: modules/file-input/test/spec.coffee [hunk 8]
```diff
@@ -0,0 +1 @@
+describe "File Input", ->
```
Comment (jmsmyth) [inline modules/file-input/test/spec.coffee]: Could do with tests once you're happy with how it works

Comment (c-frater) [inline modules/file-input/test/spec.coffee]: Yeah, I plan to add some - although unfortunately there's no way to 'set' the files to use due to security issues so it's going to be fun to test everything...

**Comment by c-frater** [comment]
@jmsmyth Pretty sure this is done now I've added it to form builder, is there anything you can see that still needs doing?

**Pull request merged**